java/src/main/kotlin/io/spine/validation/java/MessageValidationCode.kt

```diff
@@ -0,0 +1,226 @@
+    private val constraints = mutableListOf<CodeBlock>()
+
+    /**
+     * Renders the message validation code into the provided [psiFile].
+     *
+     * @param psiFile A source file with the generated Java class for the [message].
+     *
+     * @throws IllegalStateException if [psiFile] doesn't contain a Java class for the [message].
+     */
+    fun render(psiFile: PsiJavaFile) {
+        message.ruleList.forEach(::generate)
+        val messageClass = psiFile.findClass(messageClassName)
+        val builderClass = messageClass.nested("Builder")
+        execute {
+            messageClass.apply {
+                implementValidatableMessage()
+                declareValidateMethod()
+                declareSupportingFields()
+                declareSupportingMethods()
+            }
+            builderClass.apply {
+                implementValidatingBuilder()
+                injectValidationIntoBuildMethod()
+                annotateBuildReturnType()
+                annotateBuildPartialReturnType()
+            }
+        }
+    }
+
+    /**
+     * Generates a Java [constraint][constraints], [supportingFields] and [supportingMethods]
+     * for the passed [rule].
+     *
+     * A single rule always generates a single Java constraint represented by a [CodeBlock].
+     * The number of supported members is not restricted. A single rule may generate zero,
+     * one, or more supporting members.
+     */
+    private fun generate(rule: Rule) {
+        val context = newContext(rule, message)
+        val generator = generatorFor(context)
+        constraints.add(generator.code())
+        supportingFields.addAll(generator.supportingFields())
+        supportingMethods.addAll(generator.supportingMethods())
+    }
+
+    private fun newContext(rule: Rule, message: CompilationMessage) =
+        GenerationContext(
+            client = renderer,
+            typeSystem = renderer.typeSystem,
+            rule = rule,
+            msg = This(),
+            validatedType = messageType,
+            protoFile = message.type.file,
+            violationList = violations
+        )
+
+    private fun MessagePsiClass.implementValidatableMessage() {
+        val qualifiedName = ValidatableMessage::class.java.canonicalName
+        val reference = elementFactory.createInterfaceReference(qualifiedName)
+        implement(reference)
+    }
+
+    private fun MessagePsiClass.declareValidateMethod() {
+        // `CodeBlock` is printed with a new line in the end, so no need to call `joinByLine()`.
+        // And we have to trim because we don't need a trailing empty line.
+        val formattedConstraints = constraints.joinToString(separator = "").trim()
+        val psiMethod = elementFactory.createMethodFromText(
+            """
+            |public java.util.Optional<io.spine.validate.ValidationError> validate() {
+            |    ${validateMethodBody(formattedConstraints)}
+            |}
+            """.trimMargin(), this
```

Comment:
Why do we use the style which requires `trimMargin()` here, and the `trimIndent()` style below?